I18n cleanup

tfmorris opened the pull request:
### Description

This PR begins the process of repairing a decade of neglected/damage to the internationalization of OpenLibrary. It updates the HTML templates and macros, the messages.pot message catalog template which is extracted from them, the extraction program to include missing source files and directories, and finally the French translation. 

There are still some missing strings in the French translation, but it should be mostly complete and it should have all the placeholders for things which still need to be translated.

None of the other languages have been updated. These should be done separately and individually be each translator to make sure any "fuzzy" update matches are correctly resolved and all new strings are translated. The French translation includes about 1500 changes and 1000 new messages, so this will be a non-trivial task for each language.

Closes #973

### Technical

The tooling is mostly unchanged except for updates to the input file/directory list for the extraction script.

Please also review the wiki instructions for programmers to see if they make sense and match the changes that have been implemented: https://github.com/internetarchive/openlibrary/wiki/Frontend-Guide#internationalization-i18n---for-programmers

### Testing

After an initial local review, this should be place on dev where it can be reviewed by a large number of people and with enough book data to review it effectively.

Every web page in the system should be reviewed to make sure that it:
* renders without error in English and French
* has all strings translated for French

To test in French, either set your browser settings so that French is at the top of your preferred language list (note this affects all sites) or append `?lang=fr` to the URL for the page you're reviewing.

### Evidence

hornc commented on openlibrary/i18n/messages_cs.po:
```diff
@@ -1,2179 +0,0 @@
-# Translation for Open Library.
```
There are some good changes in this PR, but it has become stale though :( 

I don't think deleting this cs.po file is the right thing to do. There's not too much in it, but there is a way to auto-refresh the .po from the latest .pot which should be the process when changes are made to the .pot.

Many, but not all, of the menu strings have been translated into Czech, and will still be current: https://openlibrary.org/?lang=cs

tfmorris replied:
I don't understand the comment. It's only being deleted because it's a duplicate in the wrong place. The correct one is here: 
https://github.com/internetarchive/openlibrary/blob/09729ab073f68ca276f3fb8fb53dc77963cc3819/openlibrary/i18n/cs/messages.po

The PR will, of course, get rebased when it's finalized and no longer WIP. I can push a rebased version now, but the reason for the WIP tag is that I don't consider it ready to review yet.

tfmorris replied:
OK, it's been rebased, if that helps, but it's still WIP and not ready for review.

tfmorris replied:
@hornc Does the explanation about the cs.po file make sense? If so, can you mark this as resolved?

tfmorris replied:
@hornc Ping. I'd really like to get this merged in the next day or two so that it can't get some serious testing on dev.

hornc replied:
@tfmorris yes, makes sense -- looking back at the code it looks like this file was copied rather than moved to the correct location. I see now this PR deletes a duplicate .po file and leaves the correct one.

I hadn't realised there were 2 copies - It surprised me.

cdrini commented on openlibrary/templates/merge/authors.html:
```diff
@@ -48,33 +48,33 @@ <h1>$_("Merge Authors")</h1>
 
 <ol class="sansserif">
-    <li>Select a &quot;master&quot; record that best represents the author -
+    <li>$_('Select a &quot;master&quot; record that best represents the author -')
```
I don't think those will get rendered correctly? I think using `$` instead of `$:` will cause this to get rendered literally? But not sure.

```suggestion
    <li>$_('Select a "master" record that best represents the author -')
```

cdrini replied:
Confirmed this is the case.

cdrini commented on openlibrary/templates/merge/authors.html:
```diff
@@ -103,22 +103,22 @@ <h1>$_("Merge Authors")</h1>
                     <ul>
                     $for doc in top['books']:
-                        <li><a href="$doc.key" target="new" title="Open in a new window">$doc.title</a> <span class="smaller">$doc.edition_count edition$("s" if doc.edition_count != 1 else ""),
+                        <li><a href="$doc.key" target="new" title="$_('Open in a new window')">$doc.title</a> <span class="smaller">$ungettext('1 edition', '%(count)d editions', $doc.edition_count, doc.edition_count),
```
```suggestion
                        <li><a href="$doc.key" target="new" title="$_('Open in a new window')">$doc.title</a> <span class="smaller">$ungettext('1 edition', '%(count)d editions', doc.edition_count, count=doc.edition_count),
```

cdrini commented on openlibrary/templates/merge/authors.html:
```diff
@@ -103,22 +103,22 @@ <h1>$_("Merge Authors")</h1>
+                        <li><a href="$doc.key" target="new" title="$_('Open in a new window')">$doc.title</a> <span class="smaller">$ungettext('1 edition', '%(count)d editions', $doc.edition_count, doc.edition_count),
                         $if doc.get('first_publish_year'):
-                            <span title="First published in">$doc.first_publish_year</span>
+                            <span title="$_('First published in ')">$doc.first_publish_year</span>
```
```suggestion
                            <span title="$_('First published in')">$doc.first_publish_year</span>
```

cdrini commented on openlibrary/templates/merge/authors.html:
```diff
@@ -103,22 +103,22 @@ <h1>$_("Merge Authors")</h1>
                 </div>
                 <div class="data count">
-                    <a href="/authors/$k" target="new" title="$_('Visit this author\'s page in a new window')">$sprintf(ungettext("%s work", "%s works", top['total']), top['total'])</a>
+                    <a href="/authors/$k" target="new" title="$_('Visit this author\'s page in a new window')">$ungettext("1 work", "%(count)d works", top['total'], top['total'])</a>
```
```suggestion
                    <a href="/authors/$k" target="new" title="$_('Visit this author\'s page in a new window')">$ungettext("1 work", "%(count)d works", top['total'], count=top['total'])</a>
```

cdrini commented on openlibrary/templates/search/authors.html:
```diff
@@ -21,18 +21,18 @@
     $if q and 'error' not in results:
         $if num_found:
-            <p class="sansserif darkgreen collapse"><strong>$commify(response['numFound']) hit$("s" if response['numFound'] != 1 else "")</strong></p>
+            <p class="sansserif darkgreen collapse"><strong>$ungettext('%(count)s hit', '%(count)s hits', response['numFound'], count=commify(response['numFound']))</strong></p>
```
```suggestion
            <p class="sansserif darkgreen collapse"><strong>$ungettext('1 hit', '%(count)s hits', response['numFound'], count=commify(response['numFound']))</strong></p>
```

cdrini commented on openlibrary/templates/search/inside.html:
```diff
@@ -22,10 +22,10 @@ <h1>$_("Search Inside")</h1>
 
         $if not num_found:
-            <p class="sansserif red collapse">No hits for: <strong>$q</strong></span></p>
+            <p class="sansserif red collapse">$:_('No hits for: %(query)s', query="<strong>$q</strong></span></p>")
```
query is a user provided string; it should be displayed using `$:`, because then someone could craft a URL to run arbitrary JS on a user's computer.

cdrini replied:
Also, `$q` won't get substituted in its current position. You should be able to do something like this:

```
$def escaped_query(): <strong>$q</strong>
$:_('No hits for: %(query)s', query=escaped_query())
```

tfmorris replied:
Thanks for the suggestion. I don't think that page was actually tested, so I'll add it to the list to review.

cdrini commented on openlibrary/templates/subjects.html:
```diff
@@ -12,7 +12,7 @@ <h1 class="inline">
     <span class="heading">
         <span class="count" id="coversCount">
-            <strong><span>$sprintf(ungettext("%s work", "%s works", page.work_count), commify(page.work_count))</span></strong>
+            <strong><span>$ungettext("%(count)s work", "%(count)s works", page.work_count, count=commify(page.work_count))</span></strong>
```
```suggestion
            <strong><span>$ungettext("1 work", "%(count)s works", page.work_count, count=commify(page.work_count))</span></strong>
```

cdrini commented on openlibrary/templates/subjects.html:
```diff
@@ -2,7 +2,7 @@
 $var title: $page.name
 
-$ subject_list = [('subjects', 20), ('places', 20), ('people', 10), ('times', 10)]
+$ subject_list = [('subjects', _('SUBJECTS'), 20), ('places', _('PLACES'), 20), ('people', _('PEOPLE'), 10), ('times', _('TIMES'), 10)]
```
Uppercasing should be done in CSS; it's generally considered bad practice to put this in the text itself, since it's not related to the content, it's just a stylistic choice.

```suggestion
$ subject_list = [('subjects', _('Subjects'), 20), ('places', _('Places'), 20), ('people', _('People'), 10), ('times', _('Times'), 10)]
```

cdrini commented on openlibrary/templates/subjects.html:
```diff
@@ -73,11 +73,11 @@ <h2>$_("Related...")
-          <h6 class="black collapse uppercase">$s</h6>
-          <div class="subjects" id="subjects-$s">
-            $:renderSubjects(page[s][:limit])
+          <h6 class="black collapse">$label</h6>
```
```suggestion
          <h6 class="black collapse uppercase">$label</h6>
```

cdrini commented on openlibrary/templates/type/author/view.html:
```diff
@@ -71,7 +71,7 @@
     $:macros.databarView(page)
 
-    <h1>$page.get('name', 'name missing')</h1>
+    <h1>$page.get('name', _('name missing'))</h1>
```
```suggestion
    <h1>$title</h1>
```

cdrini commented on openlibrary/templates/type/author/view.html:
```diff
@@ -129,30 +129,27 @@ <h6 class="collapse black uppercase">$_("Location")</h6>
                 <h2 class="collapse">
-                    $commify(books.num_found) work$("s" if books.num_found != 1 else "")
-                    <span class="count smaller"><a href="/books/add?author=$page.key">Add another</a>?</span>
+                    $ungettext("%(count)d work", "%(count)d works", books.num_found, count=books.num_found)
```
```suggestion
                    $ungettext("1 work", "%(count)s works", books.num_found, count=commify(books.num_found))
```

cdrini commented on openlibrary/templates/type/edition/view.html:
```diff
@@ -216,24 +219,26 @@ <h2 class="edition-byline">$_("by") $:authors_byline</h2>
                         <span class="adjust">.</span>
                     $if page.languages:
-                    <br/>Written in <span itemprop="inLanguage">$:thingrepr(page.languages)</span>.
+                        <br/>$:_('Written in')
```
```suggestion
                        $def lang_span: <span itemprop="inLanguage">$:thingrepr(page.languages)</span>
                        <br/>$:_('Written in %(lang)s.', lang=str(lang_span()))
```

cdrini commented on openlibrary/templates/type/list/embed.html:
```diff
@@ -101,9 +100,9 @@ <h3>
                         <span class="type people small">$seed.type</span>
                     $elif seed.type == "work":
-                        - $commify(seed.edition_count) edition$("s" if seed.edition_count != 1 else "")$cond(seed.ebook_count, ",")
+                        $ungettext('- %(count) edition', '- %(count)s editions', seed.edition_count,  count=commify(seed.edition_count))$cond(seed.ebook_count, ",")
```
```suggestion
                        $ungettext('- 1 edition', '- %(count)s editions', seed.edition_count,  count=commify(seed.edition_count))$cond(seed.ebook_count, ",")
```

cdrini commented on openlibrary/templates/type/list/embed.html:
```diff
@@ -101,9 +100,9 @@ <h3>
+                        $ungettext('- %(count) edition', '- %(count)s editions', seed.edition_count,  count=commify(seed.edition_count))$cond(seed.ebook_count, ",")
                     $else:
-                        - $commify(seed.work_count) work$("s" if seed.work_count != 1 else "")$cond(seed.ebook_count, ",")
+                        $ungettext('- %(count)s work', '- %(count)s works', seed.work_count, count=$commify(seed.work_count))$cond(seed.ebook_count, ",")
```
```suggestion
                        $ungettext('- 1 work', '- %(count)s works', seed.work_count, count=$commify(seed.work_count))$cond(seed.ebook_count, ",")
```

cdrini commented on openlibrary/templates/type/list/exports.html:
```diff
@@ -5,7 +5,7 @@
             <li>
                 <span class="icon export"></span>
-                <span class="larger black"><strong>Export</strong></span><br/>
+                <span class="larger black"><strong>$('Export')</strong></span><br/>
```
```suggestion
                <span class="larger black"><strong>$_('Export')</strong></span><br/>
```

cdrini commented on openlibrary/templates/type/list/view.html:
```diff
@@ -99,7 +98,7 @@ <h1>$list.name</h1>
         resizable: false,
         buttons: {
-            "Yes, I'm sure": function() {
+            "$_('Yes, I\'m sure')": function() {
```
Note to self, test this

cdrini commented on openlibrary/templates/type/list/view.html:
```diff
@@ -208,36 +207,37 @@ <h3 class="booktitle">
                           <span class="type small">$seed.type</span>
                         $if seed.type == "edition" and seed.document.get('works'):
-                            - <a href="$seed.document.works[0].url()">$commify(work_editions_count) editions$("s" if seed.work_count != 1 else "")</a> in catalog
+                            $:ungettext('- <a href="$seed.document.works[0].url()">1 edition</a> in catalog', '- <a href="$seed.document.works[0].url()">%(count)s editions</a> in catalog', seed.work_count, count=commify(work_editions_count))
```
```suggestion
                            $:ungettext('- <a href="%(url)s">1 edition</a> in catalog', '- <a href="%(url)s">%(count)s editions</a> in catalog', seed.work_count, count=commify(work_editions_count), url=seed.document.works[0].url())
```

cdrini commented on openlibrary/templates/type/work/editions_datatable.html:
```diff
@@ -1,12 +1,12 @@
 <div class="head">
     <h2>
-    $commify(page.edition_count) $cond(page.edition_count != 1, "editions", "edition")
+    $ungettext("%(count)s edition", "%(count)s editions", page.edition_count, count = page.edition_count)
```
```suggestion
    $ungettext("1 edition", "%(count)s editions", page.edition_count, count=page.edition_count)
```

cdrini commented on openlibrary/templates/type/work/view.html:
```diff
@@ -14,26 +15,26 @@
         title = page.title
         if page.subtitle:
-             title += ": " + page.subtitle
+             title = _("%(title)s: %(subtitle)s", title=page.title, subtitle=page.subtitle)
```
```suggestion
             title = _("%(title)s: %(subtitle)s", title=title, subtitle=page.subtitle)
```

cdrini commented on openlibrary/templates/type/work/view.html:
```diff
@@ -14,26 +15,26 @@
 
         if page.edition_count:
-            items.append("%d %s" % (page.edition_count, cond(page.edition_count > 1, "editions", "edition")))
+            items.append(ungettext("%(count)s edition", "%(count)s editions", page.edition_count, count = page.edition_count))
```
```suggestion
            items.append(ungettext("1 edition", "%(count)s editions", page.edition_count, count=page.edition_count))
```

cdrini commented on openlibrary/templates/type/work/view.html:
```diff
@@ -63,10 +64,7 @@ <h1>
-                        <a href="#editions" class="plain brown">$page.edition_count editions</a>
-                    $else:
-                        <a href="#editions" class="plain brown">$page.edition_count edition</a>
+                    <a href="#editions" class="plain brown">$ungettext("%(count)s edition", "%(count)s editions", page.edition_count, count=page.edition_count)</a>
```
```suggestion
                    <a href="#editions" class="plain brown">$ungettext("1 edition", "%(count)s editions", page.edition_count, count=page.edition_count)</a>
```

cdrini commented on openlibrary/templates/type/work/view.html:
```diff
@@ -147,7 +145,7 @@ <h3 class="collapse">$_("Excerpts")</h3>
                                     $_("Page") $excerpt.pages,
                                 $if excerpt.author:
-                                    $_("added by") <a href="$excerpt.author.key">$excerpt.author.displayname</a>.
+                                    $_("added by %s", '<a href="$excerpt.author.key">$excerpt.author.displayname</a>.')
```
```suggestion
                                    $:_("added by %s", '<a href="$excerpt.author.key">$excerpt.author.displayname</a>.')
```

cdrini replied:
Oh, the `$` won't work here again

cdrini commented on openlibrary/templates/work_search.html:
```diff
@@ -129,30 +129,30 @@
                 $continue
             $for v in values if isinstance(values, list) else [values]:
-                <input type="hidden" name="$k" value="$v.replace('"', '&quot;')" />
+                <input type="hidden" name="$k" value="$v.replace('&quot;', '&quot;')" />
```
Isn't this a no-op now?

tfmorris replied:
Yup, my bad. emacs gets confused by the nested quotes and I messed up fixing it.

cdrini commented on openlibrary/templates/account/borrow.html:
```diff
@@ -45,7 +45,7 @@
         \$("#contentMsg").show();
         \$(".alert span").css({'font-size':'.6875em','color':'#333'})
-        \$(".alert span").text("You have 5 ebooks out at the moment. That's the limit! You can free up slots by returning books early.");
+        \$(".alert span").text($_("You have 5 ebooks out at the moment. That's the limit! You can free up slots by returning books early."));
```
```suggestion
        \$(".alert span").text("$_("You have 5 ebooks out at the moment. That's the limit! You can free up slots by returning books early.")");
```

tfmorris commented on openlibrary/macros/ReadButton.html:
```diff
@@ -1,5 +1,6 @@
 $def with(ocaid, borrow=False, listen=False, loan=None)
 
+<!-- FIXME: Not internationalizable. Can't use internal action names as user visible strings!-->
```
#2881

tfmorris commented on openlibrary/templates/home/categories.html:
```diff
@@ -19,14 +19,14 @@ <h2 class="home-h2"><a href="/subjects">Browse by Subject</a></h2>
               $if subject:
-                <p class="category-count" name="">$commify(subject.get('work_count', '')) Books</p>
+                $ count = commify(subject.get('work_count', ''))
+                <p class="category-count" name="">$ungettext('%(count)s Book', '%(count)s Books', subject.get('work_count'), count=count)</p>
```
Your proposed version has 5 parameters, but the function only takes 4 (which you listed correctly).

tfmorris commented on openlibrary/templates/home/categories.html:
```diff
@@ -19,14 +19,14 @@ <h2 class="home-h2"><a href="/subjects">Browse by Subject</a></h2>
               $if subject:
-                <p class="category-count" name="">$commify(subject.get('work_count', '')) Books</p>
+                $ count = commify(subject.get('work_count', ''))
+                <p class="category-count" name="">$ungettext('%(count)s Book', '%(count)s Books', subject.get('work_count'), count=count)</p>
```
This is missing a quote, but I'll change the four occurrences in my source tree.

tfmorris commented:
@cdrini I think I've addressed all the issues that you raised as well as testing and fixing up some of the suggested edits.

Moving to a hardcoded "1 foo" for the singular of `ungettext()` appears to have been a mistake because it renders the 0 case using this causing "0 results" to render as "1 resultat", so I'll have to back that out.

I internationalized a few more pages and also internationalized the relative datetime formatting ("24 minutes ago").

At this point, I think it's good enough to put on dev where we can get more eyes on it. Please re-review when you get a chance.

cdrini commented:
> Moving to a hardcoded "1 foo" for the singular of ungettext() appears to have been a mistake because it renders the 0 case using this causing "0 results" to render as "1 resultat", so I'll have to back that out.

Hmmm, do you by any chance know why that is? That seems like a bug.

tfmorris commented:
> Hmmm, do you by any chance know why that is? That seems like a bug.

I originally thought it was due to language specific pluralization rules, but I'm pretty sure the correct French would be 0 resultats, not 0 resultat, so I'll go back and dig into it more.

cdrini commented:
@tfmorris I think that should be specified on the language po file:

https://github.com/internetarchive/openlibrary/blob/c6940835990da9a2c5de6abb825a5ba25ebeb656/openlibrary/i18n/fr/messages.po#L21

https://github.com/internetarchive/openlibrary/blob/90227e2279b0e50a63d3f161db894f6f3c1fae00/openlibrary/i18n/es/messages.po#L17

Docs: https://www.gnu.org/software/gettext/manual/html_node/Translating-plural-forms.html

cdrini commented:
I added a test that validates the template files in #2900 ; let's get that merged in first. I found an issue like this and thought I'd save us same time by automating checking these:

https://github.com/tfmorris/openlibrary/blob/de8e313ac73aa5e0e2d0f7ee2b91860584626eee/openlibrary/macros/Pager.html#L22

```
<a href="$changequery(page=None)" class="ChoosePage">&laquo;&nbsp;$_'(First')</a>
```

cdrini commented:
Ugh, ran the test on your branch and it didn't catch the Pager.html error :( Although it did find an error on `openlibrary/templates/type/list/embed.html` line ~115

cdrini commented:
Ah, there is a bug in embed.html! phew; it's this line has an extra `$`:

```
$ungettext('- 1 work', '- %(count)s works', seed.work_count, count=$commify(seed.work_count))$cond(seed.ebook_count, ",")
```

cdrini commented:
Ah, apparently this is syntactically correct :(

```
<a href="$changequery(page=None)" class="ChoosePage">&laquo;&nbsp;$_'(First')</a>
```

tfmorris commented:
Thanks for the re-review. It sounds like you've already used your new tool to full effect, so I don't see why merging it should block this. There are many, **many** more errors which will only be found by actually testing each page individually. I've done my best, but more eyes are needed. The fact that the template compiles is no indication whether or not it will render in each language.

> @tfmorris I think that's specified in the language po file:

Yes, it is, and, as I think you're somewhat obliquely implying, it's specified incorrectly. Fixed in 6fec170e2ae1a09c5dd482924f92ca1d62fdb872. This affects the French translation only.

> Ah, there is a bug in embed.html! phew; it's this line has an extra `$`:
> 
> ```
> $ungettext('- 1 work', '- %(count)s works', seed.work_count, count=$commify(seed.work_count))$cond(seed.ebook_count, ",")
> ```

I think you're referring to openlibrary/templates/type/list/embed.html. Fixed in ba28a4c659f0ce28fc7adafc6ba597159efd708f 

> Ah, apparently this is syntactically correct :(
> 
> ```
> <a href="$changequery(page=None)" class="ChoosePage">&laquo;&nbsp;$_'(First')</a>
> ```

Yes, it is.

We're down to the point where you've spotted 1 or 2 lines wrong in a multi-thousand line patch. I'm sure there are more, but we're not going to find them like this.

@mekarpeles I'd argue that we're only going to significantly increase our rate of problem discovery by getting more eyes on the problem. ie putting it on dev/staging

My months of contribution to a topic which should be a core OpenLibrary priority are drawing to a close.

cdrini commented:
Yep! You're correct; now that we have some sort of guarantee that everything's syntactically correct, I'm happy to put this on dev :+1:

cdrini commented on openlibrary/macros/Pager.html:
```diff
@@ -19,13 +19,13 @@
     $if page != 1:
-        <a href="$changequery(page=None)" class="ChoosePage">&laquo;&nbsp;First</a>
-        <a href="$changequery(page=page-1)" class="ChoosePage">&lt;&nbsp;Previous</a>
+        <a href="$changequery(page=None)" class="ChoosePage">&laquo;&nbsp;$_'(First')</a>
```
Typo

tfmorris replied:
Fixed in 5d216dc0ff4fbdbd85d2f96dc64942dc22803e08

cdrini commented on openlibrary/templates/account/create.html:
```diff
@@ -46,12 +46,12 @@ <h1>$_("Sign Up")</h1>
                 <input type="text" class="required" name="username" id="username" value="$form.username.value" autocapitalize="off"/>
                 <span class="invalid clearfix" htmlfor="username">$form.username.note</span>
-                <div class="sansserif smaller lighter">Your URL: https://openlibrary.org/people/<span id="userUrl">screenname</span></div>
+                <div class="sansserif smaller lighter">$_('Your URL: https://openlibrary.org/people/<span id="userUrl">screenname</span>')</div>
```
Ping; we can't include `screenname` in the i18n string because it's found/replaced from the JS

tfmorris replied:
That's also true of the `<span>` tags. We've got a number of cases where formatting, etc is mixed with translatable text, as well as other strings, such as this, which shouldn't be translated. There are a couple of solutions to this:
* avoid the situation altogether (by not mixing translatable text with non translatable text)
* include translators' notes with the text to be translated
In practice we'll need both. The issue right now is that the message extractor doesn't know how to extract translators' notes comments from the code, so we'll need to include the notes out of band.

tfmorris replied:
p.s. I suggest changing the placeholder string to something more obvious to humans such as `{SCREENNAME}`

cdrini replied:
🤦‍♂ Ugh, nevermind, I completely misremembered this code. It uses the _Id_, not the contents, so it should be fine. Sorry for the confusion!

tfmorris commented:
@cdrini I switched from rebasing to merging master a few weeks ago to provide continuity for your reviews, but if you're OK with it, I'd like to rebase against master now that you've completed your initial review to get rid of all the merge commits. OK?

tfmorris commented:
@cdrini Your latest typos were fixed and the branch has been rebased. Please force refresh.
@mekarpeles You comments (not here, but in Slack) were addressed in the latest commit.

tfmorris commented:
(https://dev.openlibrary.org/authors/OL541156A/W._A._Mathieu/edit?lang=fr)
> * ❌ Test Untyped created page (https://dev.openlibrary.org/foodooo?m=edit&lang=fr)
>   
>   * 502s for French, but works for De? https://dev.openlibrary.org/foodooo?m=edit&lang=de
>   * Uses `TypeChanger.html`
>   * Previously this worked for FR: https://openlibrary.org/foodooo?m=edit&lang=fr

Fixed included in bf913dadc5bdfe21d326041620a5b24b549e09c8 (s/titre/title). This was an example of a translation-specific bug (bad variable name) which we don't currently have any automated testing for. We'd need to render every page (type) in every language to catch it.

> * ❌ Test Account Create Page https://dev.openlibrary.org/account/create?lang=fr
>   
>   * HTML visible near screename

The visible template variable? If so, fixed in 4ebc664e35ef43f055f82c22aba481cb353033b3

> * ❌ Test Lists page https://dev.openlibrary.org/lists?lang=fr
>   
>   * Lots of leaked HTML

I didn't see any HTML, but the page was largely untranslated. Translation added in bf913dadc5bdfe21d326041620a5b24b549e09c8. Let me know if there's a different page with HTML on it.

I left the link to the Californian book list, but it seems wildly inappropriate.

cdrini commented:
Reloaded on dev

- Check login in incognito: https://dev.openlibrary.org/account/create?lang=fr
- Here's what I'm seeing on lists page:
![image](https://user-images.githubusercontent.com/6251786/73204585-38b52f80-410d-11ea-9ae2-ff85b622871a.png)

tfmorris commented:
Ahh, I see, I need to create some lists to test with. But it also looks like the new message catalog didn't get compiled/loaded because all the text above is now translated.

cdrini commented:
It's working now https://dev.openlibrary.org/lists?lang=fr ; not sure why it wasn't before, but I think we can address those issues after this PR is merged.

hornc commented on openlibrary/macros/TypeChanger.html:
```diff
@@ -12,14 +12,14 @@
             <label for="type.key">$_("Page Type")</label>
             <span class="small gray">&nbsp;<a href="javascript:;"
-                aria-haspopup="true" aria-controls="typeSplain" class="slide-toggle">Huh</a>?</span>
+                aria-haspopup="true" aria-controls="typeSplain" class="slide-toggle">$_('Huh?')</a></span>
```
This kind of informal content might cause problems for translators and isn't the best in English. Not really an issue for this PR, but I think reviewing some of this content should be a separate task.

tfmorris replied:
I agree. This "folksy" style permeates the UI, as does embedded styling to **emphasize** things that don't really need emphasis, and a plethora of different ways of saying the same thing. Even if we can't eliminate all the redundant code, we should adopt consistent, standardized terminology and phrasing to reduce the amount of text which needs to be translated and make it easier to translate.

There are definitely sections where I adopted a lower effort, imperfect internationalization as a tactic to make progress on the overall goal. Even things which have been internationalized can be improved and we should review the contents of the French `messages.po` file with an eye towards improving things in a way that will reduce the work for future translations.

Now is the time to do this, before we do a second or third language.

hornc commented on openlibrary/macros/databarHistory.html:
```diff
@@ -8,12 +8,13 @@
         $ author = get_recent_author(page)
         $if author:
-            <div class="brown smaller sansserif">$_("Last edited by") <a href="$author.key" rel="nofollow">$author.displayname</a></div>
+            <div class="brown smaller sansserif">$:_('Last edited by <a href="$author.key" rel="nofollow">%(authorname)s</a>', authorname=author.displayname)</div>
```
@tfmorris this line looks wrong , tags and code enclosed by the `_(...)`

hornc replied:
discussed, and see that this was deliberate, not a typo like I first thought. Resolving.

hornc commented on openlibrary/macros/databarView.html:
```diff
@@ -8,20 +8,21 @@
             $ author = get_recent_author(page)
         $if author:
-            <div class="brown smaller sansserif">$_("Last edited by") <a href="$author.key" rel="nofollow">$author.displayname</a></div>
+            <div class="brown smaller sansserif">$:_('Last edited by <a href="$author.key" rel="nofollow">%(authorname)s</a>', authorname=author.displayname)</div>
```
this has the same problem

hornc commented on openlibrary/i18n/__init__.py:
```diff
@@ -15,7 +14,7 @@
 def _compile_translation(po, mo):
     try:
-        catalog = read_po(open(po))
+        catalog = read_po(open(po, 'rb'))
```
@tfmorris More out of curiosity, but is there a quick explanation of why this needed to change from `r` to `rb`? Just wondering in case it becomes relevant for Python 3 upgrades, or Unicode handling subtleties in future. I feel like I should understand, but I have no objection to the change.

tfmorris replied:
This is actually a Python 3 change. The related changes are all in commit f69aa934ddda34eb153780733e59a9ad5b33026e

hornc commented:
@cdrini I have reviewed the python code, and tried to get through all the content changes too. Code changes look good, I just had one qn around the binary read write change, which I feel like I should understand the background, but functionally it looks good.

cdrini commented:
Sweet! I'm nearing the end of my day here, so merging now :+1: Great work @tfmorris ! I'm sorry if the delays were annoying, but the amount of time required to review/qa a PR is at least linear to the number of lines in the PR :P

tfmorris commented:
@cdrini @hornc Thanks for your reviews! I know it was a big patch and I don't have any illusions that our combined technical reviews will be sufficient, which is why I wanted it on dev. Hopefully the community reviewers will find any significant remaining problems before it's time to deploy.

@hornc re:

> From reviewing this it seems that there are some good practice guidelines that could be extracted from this: naming variables and including them in the context, the $:_() leading colon for including html etc that could become the basis for a guide for making UI changes in future. Just linking to this PR from somewhere would be useful.

I've tried to capture the learnings in guidelines here:
* Programmers - https://github.com/internetarchive/openlibrary/wiki/Frontend-Guide#internationalization-i18n---for-programmers
* Translators - https://github.com/internetarchive/openlibrary/wiki/Frontend-Guide#internationalization-i18n---for-translators

Let me know if you see anything missing. I may split this out into a separate page and definitely want to include some more higher level conceptual / life cycle documentation.
